Source/WebCore/dom/Document.cpp

```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
     return nullptr;
 }
 
+HTMLElement* Document::topmostAutoPopover() const
+{
+    for (auto& element : makeReversedRange(m_topLayerElements)) {
+        if (auto* candidate = downcast<HTMLElement>(element.ptr()); candidate->popoverState() == PopoverState::Auto)
+            return candidate;
+    }
+
+    return nullptr;
+}
+
+// https://html.spec.whatwg.org/#hide-all-popovers-until
+void Document::hideAllPopoversUntil(Element* endpoint, FocusPreviousElement focusPreviousElement, FireEvents fireEvents)
+{
+    Vector<Ref<HTMLElement>> popoversToHide;
+    for (auto& element : makeReversedRange(m_topLayerElements)) {
+        if (!is<HTMLElement>(element) || downcast<HTMLElement>(element.get()).popoverState() != PopoverState::Auto)
+            continue;
+        if (element.ptr() == endpoint)
+            break;
+        popoversToHide.append(downcast<HTMLElement>(element.get()));
```

Comment:
Seems like an unnecessarily awkward idiom.
```suggestion
        if (!is<HTMLElement>(element))
            continue;
        auto& popover = downcast<HTMLElement>(element.get());
        if (popover.popoverState() != PopoverState::Auto || &popover == endpoint)
            break;
        popoversToHide.append(popover);
```

Reply:
The suggested code isn't quite correct, since we want to keep iterating if the popoverState isn't auto.

`auto& popover = downcast<HTMLElement>(element.get());` also isn't quite correct, because a html element in top layer could also be a dialog element, or a fullscreen element, not necessarily a popover (so the popover state, as the spec calls it, would be none).

Reply:
Changed it to this, which avoids the downcasting twice:

```
    Vector<Ref<HTMLElement>> popoversToHide;
    for (auto& item : makeReversedRange(m_topLayerElements)) {
        if (!is<HTMLElement>(item))
            continue;
        auto& element = downcast<HTMLElement>(item.get());
        if (element.popoverState() == PopoverState::Auto) {
            if (&element == endpoint)
                break;
            popoversToHide.append(element);
        }
    }
```